psc-flink/src/main/java/com/pinterest/flink/streaming/connectors/psc/table/PscRateLimitMap.java

```diff
@@ -0,0 +1,120 @@
+ * emitting metrics to track throttling behavior.
+ *
+ *
+ *
+ * @param <T> The type of records flowing through this map function
+ */
+public class PscRateLimitMap<T> extends RichMapFunction<T, T> {
+
+    private static final long serialVersionUID = 1L;
+
+    private final double rateLimitRecordsPerSecond;
+    private transient Counter throttleOccurredCounter;
+    private transient Gauge<Long> maxThrottleDelayGauge;
+    private transient RateLimiter subtaskRateLimiter;
+    private volatile long maxThrottleDelayMs = 0;
+
+    /**
+     * Creates a new rate limiting map function.
+     *
+     * @param rateLimitRecordsPerSecond 
+     */
+    public PscRateLimitMap(double rateLimitRecordsPerSecond) {
+        this.rateLimitRecordsPerSecond = rateLimitRecordsPerSecond;
+    }
+
+    @Override
+    public void open(Configuration configuration) throws Exception {
+        super.open(configuration);
+
+        // Register metrics
+        throttleOccurredCounter = getRuntimeContext()
+                .getMetricGroup()
+                .counter("throttleOccurred");
+
+        maxThrottleDelayGauge = getRuntimeContext()
+                .getMetricGroup()
+                .gauge("maxThrottleDelayMs", () -> maxThrottleDelayMs);
+
+        // Calculate per-subtask rate limit
+        int numberOfParallelSubtasks = getRuntimeContext().getNumberOfParallelSubtasks();
+        double subtaskRateLimit = rateLimitRecordsPerSecond / numberOfParallelSubtasks;
+
+        Preconditions.checkArgument(
+                subtaskRateLimit > 0.1,
+                "Subtask rate limit should be greater than 0.1 QPS. " +
+                        "Current rate: %s records/second divided by %s subtasks = %s records/second per subtask. " +
+                        "Consider increasing the rate limit or decreasing parallelism.",
+                rateLimitRecordsPerSecond,
+                numberOfParallelSubtasks,
+                subtaskRateLimit);
+
+        subtaskRateLimiter = RateLimiter.create(subtaskRateLimit);
+    }
+
+    @Override
+    public T map(T record) throws Exception {
+        blockIfNecessary();
+        return record;
+    }
+
+    /**
+     * Blocks if necessary to enforce the rate limit.
+     * Emits metrics when throttling occurs.
+     */
+    private void blockIfNecessary() {
+        // tryAcquire returns false if there are no permits available at this time
+        if (!subtaskRateLimiter.tryAcquire()) {
+            // Record that throttling occurred
+            throttleOccurredCounter.inc();
+            long startTime = System.currentTimeMillis();
+
+            // Block until a permit becomes available
+            subtaskRateLimiter.acquire();
+
+            long endTime = System.currentTimeMillis();
+            long delayMs = endTime - startTime;
+            
+            // Update max throttle delay for monitoring
+            if (delayMs > maxThrottleDelayMs) {
```

Comment:
why only track the `maxThrottleDelayMs`? Wondering if it makes sense to just emit the latest `delayMs`. otherwise we can only see updated metrics here if it exceeds the max.

Reply:
@KevBrowne, how often do we report this metrics? Is it ever going to be reset periodically?

@jeffxiang are you proposing to have a histogram metrics capturing all `delayMs` instead of only reporting the `maxThrottleDelayMs`?